Implementing Shor's algorithm using the Qrisp framework on Cat Qubits.ipynb

```diff
@@ -0,0 +1,770 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ce4d5d33",
+   "metadata": {},
+   "source": [
+    "#### Quantum Period finding\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "9662c206-a627-4a79-b3e8-0abe336f9140",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sympy import continued_fraction_convergents, continued_fraction_iterator, Rational\n",
+    "\n",
+    "def get_r_candidates(approx):\n",
+    "    rationals = continued_fraction_convergents(continued_fraction_iterator(Rational(approx)))\n",
+    "    return [rat.q for rat in rationals]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "81e211ed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_phase_candidates(a, N):\n",
+    "    qg = QuantumModulus(N)\n",
+    "    qg[:] = 1\n",
+    "    qpe_res = QuantumFloat(2*qg.size + 1, exponent = -(2*qg.size + 1))\n",
+    "    h(qpe_res)\n",
+    "    x=a\n",
+    "    for i in range(len(qpe_res)):\n",
+    "        with control(qpe_res[i]):\n",
+    "            qg *= x\n",
+    "            x = (x*x)%N\n",
+    "    QFT(qpe_res, inv = True)\n",
+    "    qiskit_qc = qpe_res.qs.compile().to_qiskit() # compile the circuit to qiskit for technical limitations\n",
+    "                                                 # of the qrisp framework\n",
+    "    cr = ClassicalRegister(len(qpe_res))\n",
+    "    qiskit_qc.add_register(cr)\n",
+    "    qiskit_qc.measure(range(len(qg),len(qpe_res)+len(qg)),range(len(cr)))\n",
+    "    results = execute(qiskit_qc, backend=backend, shots=5).result().get_counts()\n",
+    "    return results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "5c03bb89",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def binary_to_float(binaries):\n",
+    "    # Convert binary string to floating-point decimal\n",
+    "    results = []\n",
+    "    for binary in binaries:\n",
+    "        decimal = 0\n",
+    "        for i, bit in enumerate(binary):\n",
+    "            if bit == '1':\n",
+    "                decimal += 2 ** -(i + 1)\n",
+    "        results.append(decimal)\n",
+    "    return results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "6934c408",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def quantum_shor(a,N):\n",
+    "    meas_res = get_phase_candidates(a,N)\n",
+    "    r_candidates = sum([get_r_candidates(approx) for approx in binary_to_float(meas_res.keys())], [])\n",
+    "    for cand in sorted(set(r_candidates)):\n",
+    "        if (a**cand)%N == 1:\n",
+    "            r = cand\n",
+    "            break\n",
+    "    else:\n",
+    "        raise Exception(\"Please sample again\")\n",
```

Comment:
We could add more details to the error about why we should sample again for example. Same thing for the next exception below

Reply:
Well, I'm not sure about that. These are technicalities of the Shor's algorithm and not of the implementation. If we explain the exceptions than we should explain Shor in more details, but that is not the point

Reply:
First exception could be: "Could not find the right candidate for the period. Please try resampling"

Second: "Found an odd period for the given base. Please retry with another value of a"